New model reg revamp (#523)
wandb/docs: docs/guides/model_registry/consume-models.md

---
description: ''
displayed_sidebar: default
---

> # Download a model version

Reviewer: @noaleetz Do we want to add downloading files via the UI here?

Reviewer 2: Yes - I think that is a great idea so folks know how they can grab individual model files. @ngrayluna maybe we can add a short gif to show people how to download individual files from the Files Browser. 

https://github.com/wandb/docodile/assets/40642416/0b48fe4b-f9f2-4083-8dcd-2a2fbe412c79

> Use the W&B Python SDK to download a model that you have linked to the Model Registry. This is particularly useful if you want to evaluate its performance, make predictions against a dataset, or use in a live production context. 

Reviewer 2: @ibindlish - does use_artifact call download() under the hood?

Reviewer: No. `use_model()` calls `use_artifact()` and then `download()` under the hood.

# Access and download model. Returns path to downloaded artifact
downloaded_model_path = run.use_model(name="<your-model-name>")
```

Reference a model version with one of following formats listed:

* `latest` - which will fetch the most recently linked Version
* `v#` - using `v0`, `v1`, `v2`, ... you can fetch a specific version in the Registered Model
* `alias` - specify the custom alias that you and your team assigned to your model version

See [`use_model`](../../ref/python/run.md#use_model) in the API Reference guide for more information on possible parameters and return type.

<details>

<summary>Example: Download and use a logged model</summary>

For example, in the proceeding code snippet a user called the `use_model` API. They specified the name of the model artifact they want to fetch and they also provided a version/alias. They then stored the path that is returned from the API to the `downloaded_model_path` variable.

```python
import wandb

entity = "luka"
project = "NLP_Experiments"
alias = "latest"  # semantic nickname or identifier for the model version
model_artifact_name = "fine-tuned-model"

# Initialize a run

> run = wandb.init(project=project, entity=entity)

Reviewer: ```suggestion
run = wandb.init()
```

Reviewer: context: there's a known sdk bug due to which the `use_artifact()` call doesn't work as expected if you init the run with the entity and project name :(

# Access and download model. Returns path to downloaded artifact

> downloaded_model_path = run.use_model(name=f"{model_artifact_name}:{alias}")

Reviewer: ```suggestion
downloaded_model_path = run.use_model(name=f"{entity/project/model_artifact_name}:{alias}")
```

Reviewer 2: @ibindlish do we need to include the model path here (curious what you motivation is for adding it)?

Reviewer: Mentioned in my comment above. https://github.com/wandb/docodile/pull/523#discussion_r1442433540

---
description: ''
displayed_sidebar: default
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Download a model version

> Use the W&B Python SDK or the W&B App UI to download a model that you have linked to the Model Registry. Downloading a model is particularly useful if you want to evaluate a model's performance, make predictions with a dataset, or use ship the model to production.

Reviewer 2: ```suggestion
Use the W&B Python SDK to download a model artifact that you have linked to the Model Registry. Downloading a model is particularly useful if you want to load or consume a model in a future run in order to evaluate a model's performance, make predictions with a dataset, or use ship the model to production. 

You can download individual files in an artifact in the Model Registry UI. 
```

You can download individual files but not the whole artifact. Typically the main reason why people download is so that they can have it locally in order to use in a later run.

Author: This do mentions downloading the model itself, not the entire artifact...?

Reviewer 2: I think we should only refer to logging a model artifact (i.e the full directory of files which can be multiple, not just the weights file) as this whole package is what people will be downloading and loading into their run. Because downloading in the UI and SDK don't both act on 1 object (i.e download() is called on whole artifact while UI you download individual files) we should focus on download() via SDK and shouldn't advertise them as identical actions. Does that make sense? 

i.e I think we should should discuss the SDK version as the primary way to download, and we can add a note at the end that says you can also download individual files in the file browser (but not the whole directory). 

We can include but I just want to make sure it is clear that the browser is more for downloading individual files, not for loading the whole directory onto your machine.

Author: Ah. I see what you mean. Makes sense.

:::info
You are responsible for providing additional Python functions, API calls to reconstruct, deserialize your model into a form that you can work with. 

W&B suggests that you document information on how to load models into memory with model cards. For more information, see the [Document machine learning models](./create-model-cards.md) page. 
:::

<Tabs
  defaultValue="cli"
  values={[
    {label: 'CLI', value: 'cli'},
    {label: 'W&B App', value: 'app'},
  ]}>
  <TabItem value="cli">

Replace values within `<>` with your own:

```python
import wandb

# Initialize a run
run = wandb.init(project="<project>", entity="<entity>")

# Access and download model. Returns path to downloaded artifact
downloaded_model_path = run.use_model(name="<your-model-name>")
```

Reference a model version with one of following formats listed:

> * `latest` - Use `latest` tag to specify the model version that is most recently linked.

Reviewer 2: ```suggestion
* `latest` - Use `latest` alias to specify the model version that is most recently linked.
```
we have tags already for registered models so it is very important not to use tag language for aliases!
